File: lib/service/memory/memory_candidate.dart
enum MemoryDocTarget { daily, longTerm }

enum MemoryCandidateStatus { pending, applied, dismissed }

class MemoryCandidate {
  const MemoryCandidate({
    required this.id,
    required this.summary,
    required this.text,
    required this.targetDoc,
    required this.sourceType,
    required this.createdAtMs,
    required this.status,
    this.conversationId,
    this.messageNodeId,
    this.sensitivity = 'normal',
    this.confidence,
    this.appliedAtMs,
    this.displayText,
    this.sourcePointer,
    this.rawContextRef,
    this.appliedTargetDoc,
    this.reviewedAtMs,
    this.dismissedAtMs,
    this.decisionSource,
    this.triggerKind,
  });

  final String id;
  final String summary;
  final String text;

  /// The system's original recommendation.
  final MemoryDocTarget targetDoc;

  /// The destination actually chosen when the candidate is applied.
  final MemoryDocTarget? appliedTargetDoc;

  final String sourceType;
  final int createdAtMs;
  final MemoryCandidateStatus status;
  final String? conversationId;
  final String? messageNodeId;
  final String sensitivity;
  final double? confidence;
  final int? appliedAtMs;
  final int? reviewedAtMs;
  final int? dismissedAtMs;
  final String? decisionSource;
  final String? triggerKind;
  final String? displayText;
  final String? sourcePointer;
  final String? rawContextRef;

  MemoryDocTarget get effectiveTargetDoc => appliedTargetDoc ?? targetDoc;

  bool get isPending => status == MemoryCandidateStatus.pending;
  bool get isApplied => status == MemoryCandidateStatus.applied;
  bool get isDismissed => status == MemoryCandidateStatus.dismissed;

  String get effectiveDisplayText {
    final value = (displayText ?? '').trim();
    if (value.isNotEmpty) {
      return value;
    }
    return text;
  }

  String get effectiveSourcePointer {
    final value = (sourcePointer ?? '').trim();
    if (value.isNotEmpty) {
      return value;
    }
    final conversation = (conversationId ?? '').trim();
    final messageNode = (messageNodeId ?? '').trim();
    if (conversation.isEmpty && messageNode.isEmpty) {
      return '';
    }
    if (conversation.isNotEmpty && messageNode.isNotEmpty) {
      return '$conversation#$messageNode';
    }
    return conversation.isNotEmpty ? conversation : messageNode;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'summary': summary,
      'text': text,
      'targetDoc': targetDoc.name,
      'appliedTargetDoc': appliedTargetDoc?.name,
      'sourceType': sourceType,
      'createdAtMs': createdAtMs,
      'status': status.name,
      'conversationId': conversationId,
      'messageNodeId': messageNodeId,
      'sensitivity': sensitivity,
      'confidence': confidence,
      'appliedAtMs': appliedAtMs,
      'reviewedAtMs': reviewedAtMs,
      'dismissedAtMs': dismissedAtMs,
      'decisionSource': decisionSource,
      'triggerKind': triggerKind,
      'displayText': displayText,
      'sourcePointer': sourcePointer,
      'rawContextRef': rawContextRef,
    };
  }

  factory MemoryCandidate.fromJson(Map<String, dynamic> json) {
    return MemoryCandidate(
      id: (json['id'] ?? '').toString(),
      summary: (json['summary'] ?? '').toString(),
      text: (json['text'] ?? '').toString(),
      targetDoc: MemoryDocTarget.values.firstWhere(
        (value) =>
            value.name == (json['targetDoc'] ?? json['proposedTargetDoc']),
        orElse: () => MemoryDocTarget.daily,
      ),
      appliedTargetDoc: () {
        for (final value in MemoryDocTarget.values) {
          if (value.name == json['appliedTargetDoc']) {
            return value;
          }
        }
        return null;
      }(),
      sourceType: (json['sourceType'] ?? '').toString(),
      createdAtMs: (json['createdAtMs'] as num?)?.toInt() ?? 0,
      status: MemoryCandidateStatus.values.firstWhere(
        (value) => value.name == json['status'],
        orElse: () => MemoryCandidateStatus.pending,
      ),
      conversationId: json['conversationId']?.toString(),
      messageNodeId: json['messageNodeId']?.toString(),
      sensitivity: (json['sensitivity'] ?? 'normal').toString(),
      confidence: (json['confidence'] as num?)?.toDouble(),
      appliedAtMs: (json['appliedAtMs'] as num?)?.toInt(),
      reviewedAtMs: (json['reviewedAtMs'] as num?)?.toInt(),
      dismissedAtMs: (json['dismissedAtMs'] as num?)?.toInt(),
      decisionSource: json['decisionSource']?.toString(),
      triggerKind: json['triggerKind']?.toString(),
      displayText: json['displayText']?.toString(),
      sourcePointer: json['sourcePointer']?.toString(),
      rawContextRef: json['rawContextRef']?.toString(),
    );
  }

  MemoryCandidate copyWith({
    String? id,
    String? summary,
    String? text,
    MemoryDocTarget? targetDoc,
    MemoryDocTarget? appliedTargetDoc,
    String? sourceType,
    int? createdAtMs,
    MemoryCandidateStatus? status,
    String? conversationId,
    String? messageNodeId,
    String? sensitivity,
    double? confidence,
    int? appliedAtMs,
    int? reviewedAtMs,
    int? dismissedAtMs,
    String? decisionSource,
    String? triggerKind,
    String? displayText,
    String? sourcePointer,
    String? rawContextRef,
  }) {
    return MemoryCandidate(
      id: id ?? this.id,
      summary: summary ?? this.summary,
      text: text ?? this.text,
      targetDoc: targetDoc ?? this.targetDoc,
      appliedTargetDoc: appliedTargetDoc ?? this.appliedTargetDoc,
      sourceType: sourceType ?? this.sourceType,
      createdAtMs: createdAtMs ?? this.createdAtMs,
      status: status ?? this.status,
      conversationId: conversationId ?? this.conversationId,
      messageNodeId: messageNodeId ?? this.messageNodeId,
      sensitivity: sensitivity ?? this.sensitivity,
      confidence: confidence ?? this.confidence,
      appliedAtMs: appliedAtMs ?? this.appliedAtMs,
      reviewedAtMs: reviewedAtMs ?? this.reviewedAtMs,
      dismissedAtMs: dismissedAtMs ?? this.dismissedAtMs,
      decisionSource: decisionSource ?? this.decisionSource,
      triggerKind: triggerKind ?? this.triggerKind,
      displayText: displayText ?? this.displayText,
      sourcePointer: sourcePointer ?? this.sourcePointer,
      rawContextRef: rawContextRef ?? this.rawContextRef,
    );
  }
}

File: Packages/PTAIServices/Sources/PTAIServices/Memory/MemoryCandidate.swift
import Foundation

public enum MemoryDocTarget: String, Codable, CaseIterable, Sendable {
    case daily
    case longTerm
}

public enum MemoryCandidateStatus: String, Codable, CaseIterable, Sendable {
    case pending
    case applied
    case dismissed
}

public enum MemorySourceKind: String, Codable, CaseIterable, Sendable {
    case chat
    case reading
    case manual
}

public struct MemoryCandidate: Codable, Sendable, Identifiable, Equatable {
    public let id: String
    public var summary: String
    public var text: String
    public var targetDoc: MemoryDocTarget
    public var appliedTargetDoc: MemoryDocTarget?
    public var sourceType: String
    public var createdAt: Date
    public var status: MemoryCandidateStatus
    public var conversationId: String?
    public var messageNodeId: String?
    public var bookId: Int64?
    public var cfi: String?
    public var chapter: String?
    public var sourceKind: MemorySourceKind
    public var tags: [String]
    public var rationale: String?
    public var displayText: String?
    public var sourcePointer: String?
    public var rawContextRef: String?
    public var confidence: Double?
    public var triggerKind: String?
    public var appliedAt: Date?
    public var reviewedAt: Date?
    public var dismissedAt: Date?
    public var decisionSource: String?

    public init(
        id: String = UUID().uuidString,
        summary: String,
        text: String,
        targetDoc: MemoryDocTarget,
        appliedTargetDoc: MemoryDocTarget? = nil,
        sourceType: String,
        createdAt: Date = Date(),
        status: MemoryCandidateStatus = .pending,
        conversationId: String? = nil,
        messageNodeId: String? = nil,
        bookId: Int64? = nil,
        cfi: String? = nil,
        chapter: String? = nil,
        sourceKind: MemorySourceKind = .manual,
        tags: [String] = [],
        rationale: String? = nil,
        displayText: String? = nil,
        sourcePointer: String? = nil,
        rawContextRef: String? = nil,
        confidence: Double? = nil,
        triggerKind: String? = nil,
        appliedAt: Date? = nil,
        reviewedAt: Date? = nil,
        dismissedAt: Date? = nil,
        decisionSource: String? = nil
    ) {
        self.id = id
        self.summary = summary
        self.text = text
        self.targetDoc = targetDoc
        self.appliedTargetDoc = appliedTargetDoc
        self.sourceType = sourceType
        self.createdAt = createdAt
        self.status = status
        self.conversationId = conversationId
        self.messageNodeId = messageNodeId
        self.bookId = bookId
        self.cfi = cfi
        self.chapter = chapter
        self.sourceKind = sourceKind
        self.tags = tags
        self.rationale = rationale
        self.displayText = displayText
        self.sourcePointer = sourcePointer
        self.rawContextRef = rawContextRef
        self.confidence = confidence
        self.triggerKind = triggerKind
        self.appliedAt = appliedAt
        self.reviewedAt = reviewedAt
        self.dismissedAt = dismissedAt
        self.decisionSource = decisionSource
    }

    public var effectiveTargetDoc: MemoryDocTarget {
        appliedTargetDoc ?? targetDoc
    }

    public var effectiveDisplayText: String {
        if let displayText = Self.trimmed(displayText), displayText.isEmpty == false {
            return displayText
        }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    public var effectiveSourcePointer: String {
        if let sourcePointer = Self.trimmed(sourcePointer), sourcePointer.isEmpty == false {
            return sourcePointer
        }

        let conversation = Self.trimmed(conversationId)
        let message = Self.trimmed(messageNodeId)
        switch (conversation, message) {
        case let (conversation?, message?):
            return "\(conversation)#\(message)"
        case let (conversation?, nil):
            return conversation
        case let (nil, message?):
            return message
        case (nil, nil):
            return bookPointer
        }
    }

    public var bookPointer: String {
        var parts: [String] = []
        if let bookId {
            parts.append("book:\(bookId)")
        }
        if let chapter = Self.trimmed(chapter) {
            parts.append("chapter:\(chapter)")
        }
        if let cfi = Self.trimmed(cfi) {
            parts.append("cfi:\(cfi)")
        }
        return parts.joined(separator: " | ")
    }

    private static func trimmed(_ value: String?) -> String? {
        guard let value else { return nil }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

public struct MemoryDocumentSummary: Sendable, Identifiable, Equatable {
    public enum Kind: String, Codable, Sendable {
        case daily
        case longTerm
    }

    public let name: String
    public let kind: Kind
    public let preview: String
    public let modifiedAt: Date?

    public init(name: String, kind: Kind, preview: String, modifiedAt: Date?) {
        self.name = name
        self.kind = kind
        self.preview = preview
        self.modifiedAt = modifiedAt
    }

    public var id: String { name }
}
